# Project 13 : Mathematical Trading Strategies  
#### Divyansh Garg - 220377

In [6]:
# All required imports

import yfinance as yf
import pandas as pd


#### Assignment 1

Choose 5 international indices and 5 equities from yfinance, load the daily data since “2010-01-01”, and compute the Daily Returns, Cumulative Returns, Max Drawdowns, the Sharpe Ratio, and the Sortino Ratio if the stocks were bought, held for the day and sold when the market closed

Chosen international indices:  
&emsp;1. NASDAQ Composite (^IXIC)\
&emsp;2. S&P 500 (^GSPC)\
&emsp;3. Dow Jones Industrial Average (^DJI)\
&emsp;4. NYSE AMEX Composite Index (^XAX)\
&emsp;5. NYSE Composite (^NYA)\

Chosen equities:  
&emsp;1. Microsoft (MSFT)\
&emsp;2. Alphabet Inc. (GOOGL)\
&emsp;3. Apple Inc. (AAPL)\
&emsp;4. Nvidia (NVDA)\
&emsp;5. JPMorgan Chase and Co. (JPM)

In [4]:
# Download all the data for equities

eq_data = []
idx_data = []

In [12]:
eq = ['MSFT', 'GOOGL', 'AAPL', 'NVDA', 'JPM']
idx = ['^IXIC', '^GSPC', '^DJI', '^XAX', '^NYA']

In [13]:
for equity in eq:
    eq_data.append(yf.Ticker(equity).history(start="2010-01-01", end="2023-05-01"))

for index in idx:
    idx_data.append(yf.Ticker(index).history(start="2010-01-01", end="2023-05-01"))

# print(idx_data[2])

In [15]:
# Equities

for i in range(5):
    j = eq_data[i]

    # Daily Returns
    j['DailyReturnRate'] = j.apply(lambda row: (row.Close - row.Open) / row.Open, axis = 1)

    # Cumulative Returns
    crr = (j['Close'].iloc[-1] - j['Open'].iloc[0]) / j['Open'].iloc[0]
    print("Cumulative Return of ", eq[i], " is: ", crr * 100, "%")

    # Volatility - standard deviation of daily return rate annualized
    Vol = (j['DailyReturnRate'].std()) * (252 ** .5)
    print("Volatility of ", eq[i], " is: ", Vol * 100, "%")

    # Sharpe Ratio
    rfrate = 5.25
    ror = (1 + crr) ** (365 / len(j)) - 1

    sharpeRatio = (ror * 100 - rfrate) / (Vol * 100)
    print("Sharpe Ratio of ", eq[i], " is: ", sharpeRatio)

    # Sortino Ratio
    downsideVol = j[j['DailyReturnRate'] < 0]['DailyReturnRate'].std() * (252 ** 0.5)
    sortinoRatio = (ror * 100 - rfrate) / (downsideVol * 100)
    print("Sortino Ratio of ", eq[i], " is: ", sortinoRatio)

    # Max Drawdown
    print('Max Drawdown of', eq[i], 'is: ', ((j[:-1]['Low'].min() - j[:-1]['High'].max()) / j[:-1]['High'].max()) * 100, "%")

    print(j)

Cumulative Return of  MSFT  is:  1214.6492087857273 %
Volatility of  MSFT  is:  20.077485127292096 %
Sharpe Ratio of  MSFT  is:  1.3507902513441281
Sortino Ratio of  MSFT  is:  1.9520185539131092
Max Drawdown of MSFT is:  -94.93300822218829 %
                                 Open        High         Low       Close   
Date                                                                        
2010-01-04 00:00:00-05:00   23.321034   23.686614   23.298184   23.572371  \
2010-01-05 00:00:00-05:00   23.496210   23.686617   23.336268   23.579988   
2010-01-06 00:00:00-05:00   23.519056   23.671382   23.244871   23.435278   
2010-01-07 00:00:00-05:00   23.328645   23.381960   22.993530   23.191553   
2010-01-08 00:00:00-05:00   23.062076   23.519051   23.031610   23.351494   
...                               ...         ...         ...         ...   
2023-04-24 00:00:00-04:00  281.474674  284.328452  278.112030  281.155365   
2023-04-25 00:00:00-04:00  278.900319  280.985757  274.769335  2

In [16]:
# World Indices

for i in range(5):
	j = idx_data[i]

	# Daily Returns
	j['DailyReturnRate'] = j.apply(lambda row: (row.Close - row.Open) / row.Open, axis = 1)

	# Cumulative Returns
	crr = (j['Close'].iloc[-1] - j['Open'].iloc[0]) / j['Open'].iloc[0]
	print("Cumulative Return of ", idx[i], " is: ", crr * 100, "%")

	# Volatility - standard deviation of daily return rate annualized
	Vol = (j['DailyReturnRate'].std()) * (252 ** .5)
	print("Volatility of ", idx[i], " is: ", Vol * 100, "%")

	# Sharpe Ratio
	rfrate = 5.25
	ror = (1 + crr) ** (365 / len(j)) - 1

	sharpeRatio = (ror * 100 - rfrate) / (Vol * 100)
	print("Sharpe Ratio of ", idx[i], " is: ", sharpeRatio)

	# Sortino Ratio
	downsideVol = j[j['DailyReturnRate'] < 0]['DailyReturnRate'].std() * (252 ** 0.5)
	sortinoRatio = (ror * 100 - rfrate) / (downsideVol * 100)
	print("Sortino Ratio of ", idx[i], " is: ", sortinoRatio)

	# Max Drawdown
	print('Max Drawdown of', idx[i], 'is: ', ((j[:-1]['Low'].min() - j[:-1]['High'].max()) / j[:-1]['High'].max()) * 100, "%")

	print(j)

Cumulative Return of  ^IXIC  is:  432.88560224552225 %
Volatility of  ^IXIC  is:  16.034357851986737 %
Sharpe Ratio of  ^IXIC  is:  0.9184951082246259
Sortino Ratio of  ^IXIC  is:  1.2085574331546296
Max Drawdown of ^IXIC is:  -87.28651250948418 %
                                   Open          High           Low   
Date                                                                  
2010-01-04 00:00:00-05:00   2294.409912   2311.149902   2294.409912  \
2010-01-05 00:00:00-05:00   2307.270020   2313.729980   2295.620117   
2010-01-06 00:00:00-05:00   2307.709961   2314.070068   2295.679932   
2010-01-07 00:00:00-05:00   2298.090088   2301.300049   2285.219971   
2010-01-08 00:00:00-05:00   2292.239990   2317.600098   2290.610107   
...                                 ...           ...           ...   
2023-04-24 00:00:00-04:00  12053.469727  12103.580078  11960.299805   
2023-04-25 00:00:00-04:00  11968.809570  11990.459961  11798.769531   
2023-04-26 00:00:00-04:00  11913.230469  1

## World Indices

||IXIC|GSPC|DJI|XAX|NYA|
|:---:|:----:|:----:|:----:|:----:|:----:|
|Cumulative Returns|432.885%|273.421%|226.902%|130.115%|116.366%|
|Volatility|16.034%|14.863%|14.489%|17.269%|15.613%|
|Sharpe Ratio|0.810|0.347|0.468|0.070|0.046|
|Sortino Ratio|1.066|0.438|0.599|0.090|0.054|
|Max Drawdown|-87.286%|-79.020%|-73.982%|-72.986%|-63.561%|

## Equities

||MSFT|GOOGL|AAPL|NVDA|JPM|
|:---:|:----:|:----:|:----:|:----:|:----:|
|Cumulative Returns|1214.649%|584.154%|2515.431%|6435.053%|368.125%|
|Volatility|20.077%|20.506%|22.243%|36.594%|21.322%|
|Sharpe Ratio|1.350|0.879|1.681|1.431|0.611|
|Sortino Ratio|1.952|1.237|2.480|2.169|0.885|
|Max Drawdown|-94.933%|-92.839%|-96.815%|-99.426%|-87.778%|